File: C_Coloring_Game.cpp
#include <bits/stdc++.h>
using namespace std;

int countInRange(const vector<int>& arr, int l, int r) {
    if (l > r) return 0;
    auto left = lower_bound(arr.begin(), arr.end(), l);
    auto right = upper_bound(arr.begin(), arr.end(), r);
    return (int)(right - left);
}


int main() {
    cin.tie(nullptr);
    ios::sync_with_stdio(false);
    int t; cin >> t;
    while (t--) {
        int n; cin >> n;
        vector<int> A(n); for (int i = 0; i < n; i++) cin >> A[i];
        sort(A.begin(), A.end());
        unordered_map<int,int> c;
        for (auto x : A) c[x]++;
        long long res = 0;
        int mx = *max_element(A.begin(), A.end());
        for (int i = 0; i < A.size(); i++) {
            int a = A[i];
            for (int j = i + 1; j < A.size(); j++) {
                int b = A[j];
                int maxC = a + b - 1;
                int minC = mx - a - b + 1;
                minC = max(minC, b + 1);
                int cnt = countInRange(A, minC, maxC);
                res += cnt;
            }
        }
        // add triplets
        for (const auto& kv : c) {
            auto key = kv.first;
            if (c[key] >= 3) {
                if (key * 3 > mx) {
                    long long ways = 1LL * c[key] * (c[key] - 1) * (c[key] - 2) / 6;
                    res += ways;
                }
            }
        }

        // add pairs
        vector<long long> keys;
        keys.reserve(c.size());
        for (auto &kv : c) keys.push_back(kv.first);
        sort(keys.begin(), keys.end());
    
        for (int i = 0; i < (int)keys.size(); i++) {
            long long low = keys[i];
            for (int j = i + 1; j < (int)keys.size(); j++) {
                long long high = keys[j];
                if (c[high] >= 2 && high + high + low > mx) {
                    res += c[low] * c[high] * (c[high] - 1) / 2;
                }
            }
        }

        cout << res << endl;
    }
}

File: C_Coloring_Game.py
import bisect
from collections import Counter

def countInRange(arr, l, r):
    if l > r:
        return 0
    left = bisect.bisect_left(arr, l)
    right = bisect.bisect_right(arr, r)
    return right - left

def nc3(n):
    return int(n * (n-1) * (n-2) / 6)

def nc2(n):
    return int(n * (n - 1) / 2)
fmax = lambda x, y: x if x > y else y
def solve(A):
    # print('=========')
    # print(f'{A=}')
    A.sort()
    c = Counter(A)
    res = 0
    mx = A[-1]
    for i in range(len(A)):
        a = A[i]
        for j in range(i + 1, len(A) - 1):
            # print(f'{i=}, {j=}')
            b = A[j]
            maxC = a + b - 1
            minC = mx - a - b + 1
            minC = fmax(minC, b + 1) # only take elements > b, we will handle b separately
            # print(f'minC: {minC} maxC: {maxC}')
            cnt = countInRange(A, minC, maxC)
            # print(f'count: {cnt}')
            res += cnt

            # if a + b + b > mx:
            #     print(f'could take a b though')
            #     res += c[b]
            #     print(f'count of that is: {c[b]}')
            #     res -= 1
            #     if a == b:
            #         res -= 1
            #     print(f'res now: {res}')

    # any triple that sums to > mx works
    for key in c.keys():
        if c[key] >= 3:
            if key * 3 > mx:
                ways = nc3(c[key])
                res += ways
    
    # print(f'res after triplets: {res}')

    keys = sorted(c.keys())

    # any number + pair works if the sum is > mx
    for i in range(len(keys)):
        low = keys[i]
        for j in range(i + 1, len(keys)):
            # pick 1 low, 2 high
            high = keys[j]
            if c[high] >= 2 and high + high + low > mx:
                ways = c[low] * nc2(c[high])
                res += ways


            

    return res


# Iterate on the smallest 2 numbers I pick, A and B
# I cannot take a C >= A + B or Bob will color that
# If I take a smaller C, Bob can take the max, so I need a C such that A + B + C > MAX

# max C I can take: A + B - 1
# min C I can take: MAX - A - B + 1
# min C must also be >= B

t = int(input())
for _ in range(t):
    n = int(input())
    A = list(map(int, input().split()))
    print(solve(A))





# import random

# def genRandomArray(size, low, high):
#     return sorted([random.randint(low, high) for _ in range(size)])


# def brute(A):
#     res = 0
#     mx = max(A)
#     for i in range(len(A)):
#         for j in range(i+1, len(A)):
#             for k in range(j+1, len(A)):
#                 if A[i] + A[j] + A[k] > mx:
#                     if A[k] < A[i] + A[j]:
#                         res += 1
#     return res


# for _ in range(100):
#     r = genRandomArray(100, 1, 20)
#     b = brute(r)
#     my = solve(r)
#     if b != my:
#         print("FAIL")